### ISSUE TITLE
Implementation of MDN Grizzly aka MDN "Snippets" aka MDN Web DNA CTA

### ISSUE THREAD

**Issue by schalkneethling** [mdn/sprints#1092]
# User story

After having a discussion and follow up back and forth with @atopal during the last sprint, see https://github.com/mdn/sprints/issues/1037

It felt more and more like what we need is a functionality that is a blend of the [Firefox snippet service](https://github.com/mozmeao/snippets-service), and [Django soapbox](https://pypi.org/project/django-soapbox/) without the Django admin area.

### Needs that has been defined

- Should be able to update CTAs or add new items without needing a Kuma production push
- Target a specific set of pages, for example, https://developer.mozilla.org/docs/Web/CSS/
- Target a specific page, for example, https://developer.mozilla.org/
- Use a combination of HTML, CSS and JavaScript
- Set a start and/or end date

### Basic functional requirements of CTA

* On page load, load CTA minimised
* CTA should be able to handle something as simple as some HTML content with a button/link up to a complex item such as the current payments banner
* When a CTA is dismissed(using the close button) it is not shown to the user again for 30 days
* If a CTA is dismissed, the next CTA in the queue should be shown on next page load
* Never show more than one CTA at a time
* Lightweight
* CTA is loaded asynchronously if there are active items

## Existing work

Based on the above information and the discussion had, I started work on a prototype. 

### Broad overview

There are two portions to the possible solution

1. MDN Grizzly

This is a very lightweight Expressjs server that responds to simple GET requests. It essentially has four end-points

1. `/get-json`
2. `/get-message/:message`
3. `/get-js/:message`
4. `/update-read-status`

### `/get-json`

The first call returns a JSON payload that is the message configuration, and looks something like this:

```
[
  {
    "name": "developer-survey",
    "js": true,
    "active": true
  }
]
```

This is a subset of the possible properties, the full list being:

* name - The name of the survey, which also maps to the relevant asset(HTML/CSS, JS) files
* js - Whether this message has a JavaScript dependency : default = false
* active - Whether this message is currently active : default = false
* target - Either "exact:<url>", "range:<url>" : default = *
** `exact` does an exact match on the URL
** `range` does an indexOf match on the URL
* startDate - The start date for the CTA : default = ""
* endDate - The end date for the CTA : default = ""

There is also a top level property called `read` which is always set to `false`.

### `/get-message/:message`

The will return the HTML and CSS for the relevant CTA. The call and URL is constructed by the JS running inside Kuma. For the above example it would be `/get-message/developer-survey`

The server will look for a file in the static directory called `developer-survey.html` and return its contents, for example:

```
<style>
  .cta-container {
    background-color: #83d0f2;
  }

  h2 {
    color: #fff;
    font-size: 2rem;
  }
</style>

<div class="cta-content-container">
  <h2 class="cta-title">
    MDN Developer Survey
  </h2>
  <p class="cta-copy">
    We want to know learn more about the needs of developers
  </p>
</div>
<a href="https://developer.mozilla.org/" class="button">Complete the survey</a>
```

On the kuma side, this will be injected into the `cta-container` on relevant page(s)

### `/get-js/:message`

If `js` was set to true in the message configuration, the server will look for a file called `developer-survey.js` and return its contents, for example:

```
(function() { console.log("I am a message"); })();
```

This will be injected as a `script` on the Kuma side after the HTML/CSS, parsed and executed by the browser.

### `/update-read-status`

This is used by the JS on the Kuma side.

### New CTA and updates

To add a new CTA or update an existing CTA the process will be similar to the current process used on interactive-examples

1. A user opens a pull request with the relevant HTML, CSS and JavaScript and/or, updates to the `messages.json` config file
2. The PR is reviewed
3. Once approved, it is merged to master
4. This will kick off a build step on Travis-CI
5. Once the successfull the new files will be copied to a location(S3?) from which MDN Grizzly serves the content

### The Kuma Side

On this side of the service we have what boils down to a message loader and queue/state manager.

The first task is to load the JSON and store it in `localStorage`. Once this is complete, the code calls `/update-read-status` which will set the top level `read` property to `true`.

#### Managing the freshness of `messages.json`

On subsequent page loads the script will:

1. Test whether the JSON is already stored locally
2. If it is, check for the `lastModified` date,
2.1 If it has been more than 24 hours, it will call the server again
2.2 If less than 24 hours, it will simply use the JSON from `localStorage`
3. If the JSON is loaded from the server(it has been more than 24 hours), it will check whether the `read` property is `false`
3.1 If it is `false`, it updates the JSON stored in localStorage as well as the `lastModified` date, and calls `/update-read-status`
3.2 If it is `true`, it simply falls back to the exising JSON in `localStorage` and only updates the `lastModified` date

#### Showing CTAs

Based on various status properties in `messages.json` such as `active`, `startDate`, `endDate`, `dismissedDate`, `target`, the queue manager will decide which message, if any, needs to be displayed to the user.

If there is a CTA to be displayed, the JS will asynchronously call MDN Grizzly for the HTML/CSS and JavaScript(if needed), inject the code into the `cta-container`, and show the CTA.

#### Base styles and JS

Inside the Kuma codebase will be base styles, as well as JavaScript for the base functionality of the CTAs. For example:

1. There is no need for each CTA to implement the minimize and dismiss functionality. The JS inside Kuma will look for these items, and if they are present attach the required event listeners.
2. Basic styling for the drawer, and general layout of items in the drawer will also be in Kuma but, can be overriden by each individual CTA as needed(should be kept to a minimum).

The base JavaScript and CSS will then only be loaded(asynchronously) if there is at least one CTA that needs to be shown(need to look into caching to ensure we load this from cache one subsequent load).

# Acceptance criteria

TBD

# Tasks

TBD

**Comment by schalkneethling** [comment]
@atopal I still need to do the acceptance criteria and tasks but, thought I would post this so you have some time to read over this. I also still need to put links to the current prototype repos.

**Comment by atopal** [comment]
Hey Schalk, 

Can you share the functionality of the prototype that you have worked on during this sprint? I wonder whether we can limit the first version of this to have value earlier, and add on as needed. 

Once acceptance criteria is that we can show this to a subset of users (say 5%), like the task completion survey
Another one is that we don't want to depend on a developer to make changes here. Since most of these are going to be similar (asking people to take part in a survey). That said, we can forgo this for the first version.

**Comment by schalkneethling** [comment]
> Can you share the functionality of the prototype that you have worked on during this sprint?

Will do.

> Once acceptance criteria is that we can show this to a subset of users (say 5%), like the task completion survey

This sounds fair enough. I am not exactly sure how this is managed from the Kuma side so, we will have to see if we need to use TrafficCop to do the segmentation.

> Another one is that we don't want to depend on a developer to make changes here. Since most of these are going to be similar (asking people to take part in a survey).

For things like this, it would be easy to do for pretty much anyone. You can simply copy and amend an existing one. This is where documentation on the project will come in.

It will still need to go through the PR, review process but, one could most likely even get by, by simply using the Github interface.

**Comment by schalkneethling** [comment]
@atopal Will probably be easiest to screen share with you tomorrow to demonstrate the basic functionality of this.

**Comment by schalkneethling** [comment]
After having talked over this with @atopal, there are some additional things to think about with regards to this. For one thing, we have a hard deadline for this, so time is a constraint.

With that in mind, here are some additional thoughts.

What we need for the hard deadline mentioned is:

1. The ability to show the Developer survey banner but,
2. be sure not to show it and the payments banner at the same time

The developer survey banner can be as simple as the following:

<img width="1678" alt="screenshot 2019-02-28 at 14 31 48" src="https://user-images.githubusercontent.com/10350960/53566701-a2c3fe00-3b65-11e9-90e4-2ac30beae4bb.png">

## Option 1

Have both the payments banner code as well as the code for the survey banner in the page. Decide which one of the two takes precedence, and show this banner on load. We then keep showing said banner until the user dismisses it.

Once the user has dismissed the banner, move on to showing the next until it is dismissed, then wait until the interval has elapsed, and start showing the first banner again.

This is pretty quick and simple, it does not allow for management of these banners, it needs developer intervention, it depends on the review process and a production push for any changes.

This is then not a long term solution but, allows us to get the pieces in place for the developer survey banner quickly.

## Option 2

Explore existing options available in the Django(admin) ecosystem that can give us the functionality we require. Essentially Django-soapbox with one or two superpowers i.e. ability to author and display richer call to action banners.

If such a solution exists, what would be the steps to get this integrated and deployed to production? Will need input here from @jwhitlock or @escattone 

## Option 3

Develop our own Django admin plugin loosely based on the work that has already been done for MDN Grizzly, but making use of the Django DB as well as the Django admin UI. Will need input from @jwhitlock and @escattone with regards to the complexity and timeline of doing something like this.

Currently, I do not foresee myself building this solution without some guidance from John and/or Ryan.

## Option 4

Complete the work on Grizzly and the messenger bits needed on the Kuma side, but pair it down for a v1

What v1 will consist of:

### MDN Grizzly

1. This will serve the config as well as content for the banners.
2. Functionality and styling for these will be kept simple by design and limited to two or three themes.
3. There will be no need for Grizzly to host CSS or JS, only the HTML portion of the banner.

### Messenger on Kuma

1. Gets JSON from Grizzly and stores it in localStorage (there will be no requirement in this version to manage freshness etc.)
2. Manage banner display status using localStorage and defined rules

### General Kuma

All styling and JavaScript will live inside the Kuma codebase

## Option 5 (dependant on option 4)

Everything in option 4, but includes all requirements as listed in the base user story.

## Option 6 (depends on option 5)

Adds a simple UI for managing banners, as well as editing and authoring banners.
The authoring UI will be simple and consist of the following

1. Name of the banner
2. A drop down where a user can select the theme
3. A field for the main headline
4. A field for the main banner copy
5. Action button copy

There will also be a view where all banners are listed, indicating whether a banner is active or not. The user will also have the ability to toggle the active state of any of the listed banners.

**Comment by schalkneethling** [comment]
This is superseded by https://github.com/mdn/sprints/issues/1130, so I am closing this. Will place a link to this issue in the Epic for later reference as needed.
